### PR TITLE
Feature/CASMCMS-7103: BOS conversion to use PCS, Deprecation of CAPMC (v2 only)

### PR DESCRIPTION
**Pull request by jsl-hpe**
## Summary and Scope

This mod contains extensive work for the replacement of CAPMC, and instead using PCS for BOS v2 only. BOS v1/BOA still use the CAPMC endpoint, and will not be converted over as it is marked for deprecation.

This primarily affects forceful power down, graceful power down, power on, and the status operators, but affects all operators indirectly that query for power state (almost all of them). As a result, extensive small TDS testing was performed to ensure operations continued smoothly.

PCS and CAPMC treat undefined power states and undefined (uninitialized entries) differently. As a result, due to periods of time when a node is unavailable, it is possible for a node to be skipped completely once the node in question reaches its maximum number of attempts. We may consider increasing the default number of attempts for power actions on a given node, while also halving the power state operator query time in order to ensure faster turn around and a greater chance to avoid undefined state.

It is worth noting that in the current release, CAPMC functionality is a direct shim to PCS already. This should give us greater confidence in ensured functionality of these associated changes.

## Issues and Related PRs

* Resolves [CASMCMS-7103](https://jira-pro.its.hpecorp.net:8443/browse/CASMCMS-7103)
* Doc change will also be needed in associated csm-docs location

### Tested on:

  * `wasp`
  * `mug`
  * Local development environment

### Test description:

Tests included full system deployment with the following BOS session operations:
- shutdown
- boot
- reboot

## Risks and Mitigations

As noted above, it was noticed that normal hardware can be listed as undefined for periods of time, especially during a fresh installation. The problem, from PCS' perspective, self-corrects,  but BOSv2 may leave this hardware in an error state indefinitely if the number of retry operations is exceeded.

## Pull Request Checklist

- [X] Version number(s) incremented, if applicable
- [X] Copyrights updated
- [X] License file intact
- [X] Target branch correct
- [X] CHANGELOG.md updated
- [X] Testing is appropriate and complete, if applicable

### CODE REVIEW

**Comment by jsl-hpe** [comment]
My intent is to squash the commits before merging, but I'll wait for approvals before that happens in case there is any feedback.

File: src/bos/operators/base.py
Comment (rbak-hpe) [inline src/bos/operators/base.py:125]: Why log the error a second time?

File: src/bos/operators/utils/clients/pcs.py
Comment (rbak-hpe) [inline src/bos/operators/utils/clients/pcs.py:206]: Does "task_deadline_minutes" actually have an impact, and if so should we tie it into the options for how long we're willing to wait for power-on/off before retries happen?

File: src/bos/operators/utils/clients/pcs.py
Comment (rbak-hpe) [inline src/bos/operators/utils/clients/pcs.py:162]: Is this currently used anywhere or just future-proofing?

File: src/bos/operators/base.py
Comment (jsl-hpe) [reply src/bos/operators/base.py:125]: In this particular case, logging the error message was insufficient for tracking down the exact location in the codebase that yielded it. When we log it as an exception, it additionally gives us the full stack location so that follow-up can occur.

File: src/bos/operators/utils/clients/pcs.py
Comment (jsl-hpe) [reply src/bos/operators/utils/clients/pcs.py:206]: It does, but there is a ton of nuance here because of how multiple operators can create requests, and actual power state is asynchronous to what has been done to act on the node. 1 is the lowest possible value that still makes sense (0 and -1 have special meaning).

File: src/bos/operators/base.py
Comment (rbak-hpe) [reply src/bos/operators/base.py:125]: You can do this without the extra message by adding "stack_info=True" as an argument to LOGGER.error

File: src/bos/operators/utils/clients/pcs.py
Comment (rbak-hpe) [reply src/bos/operators/utils/clients/pcs.py:206]: What I'm really asking is this: Our default for max_power_on_wait_time is two minutes.  Will having "task_deadline_minutes" set to 1 mean that PCS gives up after a minute and BOS will wait the extra minute for no reason?  Or does "task_deadling_minutes" not have any impact, possibly because we check the power state independently and don't care about the length of the "task".

File: src/bos/operators/base.py
Comment (rbak-hpe) [reply src/bos/operators/base.py:125]: Or just change the first call to "exception" rather than "error".  I just hate to clutter the logs with duplicate error messages.

File: src/bos/operators/utils/clients/pcs.py
Comment (jsl-hpe) [reply src/bos/operators/utils/clients/pcs.py:206]: Its the latter, and the delicate balance is that ideally requests to PCS expire faster than we make new requests so that things do not back up. This is an integer value field, meaning we can't specify .5, or a 30 second timeout. This is the lowest possible setting for PCS request lengths that we can actually use. Regardless of this, PCS is coded to "accomplish" its task after it successfully communicates the transition to the node(s) in question, where success is defined as having communicated via redfish the request to change the power state, NOT a success about if the node actually ended up doing what was requested. This is a major consideration, and also the reason we don't particularly care if the requests go through, as we're only ever interested in the actual power state of the node. Having these requests be the smallest possible value allows PCS the opportunity to instruct the node to change power states, and the immediately clears the request so that future power actions can be taken on the node without holding up subsequent action.

File: src/bos/operators/base.py
Comment (jsl-hpe) [reply src/bos/operators/base.py:125]: I had considered the latter, but determined that calling the error exceptional (we didn't or couldn't handle it, except by nature of a blank except Exception python no-no) was still valuable. It means "there is probably a bug here because something we didn't expect to happen, did." As was the case, the message about the unbound local error from existing code when PCS had an undefined power state in existing code caused me to leave the existing message in place. It isn't harmful and it also does a good job of explaining why we're seeing the stack in the first place.

File: src/bos/operators/utils/clients/pcs.py
Comment (jsl-hpe) [reply src/bos/operators/utils/clients/pcs.py:206]: follow on for clarification: PCS offers zero garuntees that a transition request is honored, only that it did due diligence to pass the message on to the redfish interface in question. The problem with increasing this time to something larger means that a request could become queued between operators (power off force, then power on graceful, for example) because they actual power state can be verified asynchronously. The concern is that multiple force power off requests can be made by the power off operator, the first one eventually "takes", the power on operator says "ok, now power on" but PCS does not honor the request because of a race condition with the forceful power off operator retrying.

In testing, keeping this value as small as possible allows the requests pipes to remain as clean as possible, especially since we only ever care what actually happens to the node.

File: src/bos/operators/base.py
Comment (rbak-hpe) [reply src/bos/operators/base.py:125]: It sounds like you're set on this, and it's not important enough to hold this up, but I will just add that "exception" isn't a log level.  LOGGER.exception logs at an ERROR level, and in fact its really just a shortcut for LOGGER.error(exec_info=True), so we're really just logging the same error at the same level twice with this.

File: src/bos/operators/utils/clients/pcs.py
Comment (rbak-hpe) [reply src/bos/operators/utils/clients/pcs.py:206]: Makes sense, thanks for the explanation.

**Review by rbak-hpe — APPROVED** [review]

File: src/bos/operators/base.py
Comment (jsl-hpe) [reply src/bos/operators/base.py:125]: Yep, you're right -- they funnel to the same function and one is just short hand for the other. I can switch it up.

**Pull request merged**
